shell/platform/android/io/flutter/plugin/common/JSONCodec.java

```diff
@@ -0,0 +1,79 @@
+package io.flutter.plugin.common;
+
+import java.nio.ByteBuffer;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.json.JSONTokener;
+
+/**
+ * A {@link MessageCodec} using UTF-8 encoded JSON messages, and supporting the same Java values
+ * as {@link JSONObject#wrap(Object)}.
+ */
+final class JSONCodec implements MessageCodec {
+    @Override
+    public ByteBuffer encodeMessage(Object message) {
+        if (message == null) {
+            return null;
+        }
+        return ByteBufferTools.extractBytes(JSONObject.wrap(message).toString());
+    }
+
+    @Override
+    public Object decodeMessage(ByteBuffer message) {
+        if (message == null) {
+            return null;
+        }
+        try {
+            final String json = ByteBufferTools.extractString(message);
+            final JSONTokener tokener = new JSONTokener(json);
+            final Object value = tokener.nextValue();
+            if (tokener.more()) {
+                throw new IllegalArgumentException("Invalid JSON");
+            }
+            return value;
+        } catch (JSONException e) {
+            throw new IllegalArgumentException("Invalid JSON", e);
+        }
+    }
+
+    @Override
+    public MethodCall decodeMethodCall(ByteBuffer message) {
```

Comment:
> Separating the plain message codec from how to encode and decode method calls and result envelopes would probably imply a two-step decoding: first decode a message, then interpret that message as a method call or an envelope (and symmetrically for encoding).

Not necessarily.  The plugin receiving the method call could know that it is using a codec that involves method calls rather than a less sophisticated code that only understands message.

To clarify, I'm not suggesting that we chain together a methodcall codec and a message codec.  I'm suggesting that some codecs are more sophisticated than others and encode/decode methodcalls rather than simply messages.